content/blog/cleaning-up-your-test-data-builders.md

```diff
@@ -0,0 +1,112 @@
+---
+title: "Cleaning Up Your Test Data Builders"
+date: 2019-03-16T13:24:49+01:00
+publishdate: 2019-03-16T13:24:49+01:00
+lastmod: 2019-03-17T00:00:00+01:00
+image: "images/blog/cleaning-up-your-test-data-builders.jpg"
+tags: [ "Cleaner Code", "Test Automation" ]
+comments: true
+draft: true
+---
+
+I still come across a lot of automated tests with many lines of code just to create an object. Even when most data is not relevant for the scenario being tested. Making the tests harder to read and maintain.
+
+Using the [Test Data Builder](http://www.natpryce.com/articles/000714.html) pattern can minimize the lines of code in your test to what's relevant. It's really improved the readability of my test automation code.
+
+I've always had the nagging feeling though that it could be better. Even when using a fairly simple builder, the code seems clunky to me most of the time.
+
+Here's my initial take on a builder to create a simple `Person` object. I only specify a name, gender and address but it's still a lot of code for just three properties.
+
+```csharp
+Person sherlock = new PersonBuilder()
+    .WithName("Sherlock Holmes")
+    .IsMan()
+    .WithAddress(
+        new AddressBuilder()
+            .WithAddressLine1("221B Baker Street")
+            .WithAddressLine2("London")
+            .Build()
+    )
+    .Build();
+```
+
+Inspired by a talk I attended a couple of months ago I decided to refactor some of my test data builders. In the rest of this post I'll go through some of the steps I've taken using the example above as the starting point.
+
+### Introducing the Object Mother
+
+Step one is to move the instantiation of the `PersonBuilder`. In the test I'm not really interested in the fact that I'm using a builder. I want 'A Person'. We can combine the [Object Mother](https://martinfowler.com/bliki/ObjectMother.html) pattern with the Test Data Builder pattern to make this happen.
+
+I've created an object mother class called `A`. It has a static `Person` property that returns a new instance of the `PersonBuilder`.  
+
+```csharp
+Person sherlock = A.Person
+    .WithName("Sherlock Holmes")
+    .IsMan()
+    .WithAddress(
+        new AddressBuilder()
+            .WithAddressLine1("221B Baker Street")
+            .WithAddressLine2("London")
+            .Build()
+    )
+    .Build();
+```
+
+### Moving a build method to the Object Mother
+
+We can take this one step further and create a `Man` property in the object mother. Combining `A.Person` and the `IsMan` method into one.
+
+```csharp
+Person sherlock = A.Man
```

Comment:
Die .Man methode, hergebruik je daar de .Person() methode? Daarom is het interessant om te laten zien hoe je Object Mother class te laten zien.